src/Service/Router/Router.php

```diff
@@ -0,0 +1,142 @@
+
+class Router
+{
+    /** @var RequestHandler\RequestHandlerInterface[][] */
+    private $requestHandlers;
+
+    /** @var UriParser\UriParser */
+    private $uriParser;
+
+    /**
+     * @param Repository\AbstractClientCredentials $clientCredentialsRepository
+     * @param Repository\AbstractToken             $tokenRepository
+     * @param Repository\AbstractUser              $userRepository
+     * @param Repository\AbstractPathInfo          $pathInfoRepository
+     *
+     * @throws \InvalidArgumentException
+     * @throws UriParser\Exception\InvalidRoute
+     */
+    public function __construct(
+        Repository\AbstractClientCredentials $clientCredentialsRepository,
+        Repository\AbstractToken $tokenRepository,
+        Repository\AbstractUser $userRepository,
+        Repository\AbstractPathInfo $pathInfoRepository
+    ) {
+        $this->uriParser = new UriParser\UriParser();
+
+        // add predefined routes
+        $this->subscribe(new Route\AuthToken(), new RequestMethod\Get(),
+            new RequestHandler\PostAuthToken(
+                $clientCredentialsRepository, $tokenRepository, $userRepository
+            )
+        );
+        $this->subscribe(new Route\V2(), new RequestMethod\Get(),
+            new RequestHandler\GetV2($clientCredentialsRepository, $pathInfoRepository)
+        );
+    }
+
+    /**
+     * @param Route\AbstractRoute                    $route
+     * @param RequestMethod\AbstractRequestMethod    $method
+     * @param RequestHandler\RequestHandlerInterface $handler
+     */
+    public function subscribe(
+        Route\AbstractRoute $route,
+        RequestMethod\AbstractRequestMethod $method,
+        RequestHandler\RequestHandlerInterface $handler
+    ) {
+        if (null === $route) {
+            throw new \InvalidArgumentException("Argument '\$route' is invalid!");
+        }
+        if (null === $method) {
+            throw new \InvalidArgumentException("Argument '\$method' is invalid!");
+        }
+        if (null === $handler) {
+            throw new \InvalidArgumentException("Argument '\$handler' is invalid!");
+        }
+
+        // the parser needs all subscribed routes to be able to parse the requested uri strings
+        $this->uriParser->addRoute($route);
+
+        // do the actual subscription by assigning a request handler to the given route and method
+        if (!array_key_exists($route->getIdentifier(), $this->requestHandlers)) {
+            $this->requestHandlers[$route->getIdentifier()] = array();
+        }
+        $this->requestHandlers[$route->getIdentifier()][(string) $method] = $handler;
+    }
+
+    /**
+     * @param Request\Request $request
+     *
+     * @return Response
+     *
+     * @throws Authenticator\Exception\Unauthorized
+     * @throws Exception\UnregisteredRoute
+     * @throws Exception\UnregisteredRouteMethod
+     * @throws RequestHandler\Exception\InvalidRequestHandler
+     * @throws UriParser\Exception\InvalidRoute
```

Comment:
getRoute() throws some more exceptions with the changes I mentioned in my previous comment:
     * @throws \RuntimeException
     * @throws \InvalidArgumentException

Reply:
Right I missed it because throws annotations are also missing in the getRoute method phpdoc. Added both.